src/model/modelTrng_No_GPU.py

```diff
@@ -0,0 +1,300 @@
+    """Replaces Keras' native ImageDataGenerator."""
+    """ Randomly select batch_size rows from the hdf5 file dataset """
+
+    #input_shape = tuple([batch_size] + list(hdf5_file['input'].attrs['lshape']) + [1])
+    input_shape = (batch_size, 32,32,32,1)
+
+    idx_master = get_idx_for_classes(hdf5_file, exclude_subset)
+
+    random_idx = get_random_idx(idx_master, batch_size)
+    imgs = hdf5_file["input"][random_idx,:]
+
+    imgs = imgs.reshape(input_shape)
+    imgs = np.swapaxes(imgs, 1,3)
+    ## Need to augment
+    #imgs = augment_data(imgs)
+
+    classes = hdf5_file["output"][random_idx, 0]
+
+    return imgs, classes
+
+def generate_data(hdf5_file, batch_size=50, subset=0, validation=False):
+    """Replaces Keras' native ImageDataGenerator."""
+    """ Randomly select batch_size rows from the hdf5 file dataset """
+
+    # If validation, then get the subset
+    # If not validation (training), then get everything but the subset.
+    if validation:
+        idx_master = get_idx_for_onesubset(hdf5_file, subset)
+    else:
+        idx_master = get_idx_for_classes(hdf5_file, subset)
+
+    # input_shape = tuple([batch_size] + list(hdf5_file['input'].attrs['lshape']) + [1])  #AL check
+    input_shape = (batch_size, 32,32,32,1)
+
+    while True:
+
+        random_idx = get_random_idx(idx_master, batch_size)
+        imgs = hdf5_file["input"][random_idx,:]
+        imgs = imgs.reshape(input_shape)
+        imgs = np.swapaxes(imgs, 1, 3)
+
+        if not validation:  # Training need augmentation. Validation does not.
+            ## Need to augment
+            imgs = augment_data(imgs)
+
+        classes = hdf5_file["output"][random_idx, 0]
+
+        yield imgs, classes
+
+
+def get_idx_for_onesubset(hdf5_file, subset=0):
+    '''
+    Get the indices for one subset to be used in testing/validation
+    '''
+
+    idx_subset = np.where( (hdf5_file["subsets"][:,0] == int(subset)) )[0]
+
+    idx = {}
+    idx[0] = np.where( (hdf5_file['output'][idx_subset,0] == 0) )[0]
+    idx[1] = np.where( (hdf5_file['output'][idx_subset,0] == 1) )[0]
+
+    return idx
+
+def plot_loss_accuracy(mdl):
+
+    H = mdl.history
+    N = np.arange(0, len(H["loss"]))
+    plt.style.use("ggplot")
+    plt.figure()
+    plt.plot(N, H["loss"], label="train_loss")
+    plt.plot(N, H["val_loss"], label="test_loss")
+    plt.plot(N, H["acc"], label="train_acc")
+    plt.plot(N, H["val_acc"], label="test_acc")
+    plt.title(" Resnet3d Model Training on Volumetric data (No GPU's) ")
+    plt.xlabel("Epoch #")
+    plt.ylabel("Loss/Accuracy")
+    plt.legend()
+    plt.savefig("no_GPU_fig.png")
+    plt.close()
+
+#### MAIN  ######
+with h5py.File(path_to_hdf5, 'r') as hdf5_file:
+
+    print("Valid hdf5 file in 'read' mode: " + str(hdf5_file))
+    file_size = os.path.getsize(path_to_hdf5)
+    print('Size of hdf5 file: {:.3f} GB'.format(file_size/2.0**30))
+
+    num_rows = hdf5_file['input'].shape[0]
```

Comment:
Do some sanity check would be better, in case invalid hdf5 file inputted.

Reply:
Checking needed datasets (e.g. input, output, subsets) in HDF5 file for sanity (else raising error).